[java] Added AvoidUncheckedExceptionsInSignatures Rule

pamidi99 opened the pull request:
<!--
Please, prefix the PR title with the language it applies to within brackets, such as *[java]* or *[apex]*. If not specific to a language, you can use *[core]*
-->

Before submitting a PR, please check that:
 - [X] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [X] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
Added a new rule SignatureDeclareThrowsRuntimeException for adding violation on methods throwing RuntimeExceptions in their signature.
**Issue** https://github.com/pmd/pmd/issues/1692

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset introduces 421 ...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 421 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 0 configuration errors.<br />
<a href="http://chunk.io/pmd/453ffc7fea5b41f7be14e20769172f13/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented on docs/pages/pmd/rules/java.md:
```diff
@@ -161,6 +161,7 @@ folder: pmd/rules
 *   [NcssTypeCount](pmd_rules_java_design.html#ncsstypecount): <span style="border-radius: 0.25em; color: #fff; padding: 0.2em 0.6em 0.3em; display: inline; background-color: #d9534f; font-size: 75%;">Deprecated</span> This rule uses the NCSS (Non-Commenting Source Statements) algorithm to determine the number of l...
 *   [NPathComplexity](pmd_rules_java_design.html#npathcomplexity): The NPath complexity of a method is the number of acyclic execution paths through that method.Whi...
 *   [SignatureDeclareThrowsException](pmd_rules_java_design.html#signaturedeclarethrowsexception): A method/constructor shouldn't explicitly throw the generic java.lang.Exception, since itis uncle...
+*   [SignatureDeclareThrowsRuntimeException](pmd_rules_java_design.html#signaturedeclarethrowsruntimeexception): A method/constructor should not explicitly declare java.lang.RuntimeException or it's subclasses ...
```
You do not need to commit the changes under `docs/pages/pmd/rules` since this is generated automatically. But it doesn't hurt either.

pamidi99 replied:
Alright! want me to undo this change?!

adangel replied:
that's not necessary, it will be overridden anyway...

adangel commented on pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/design/SignatureDeclareThrowsRuntimeExceptionTest.java:
```diff
@@ -0,0 +1,11 @@
+
+import net.sourceforge.pmd.testframework.PmdRuleTst;
+
+public class SignatureDeclareThrowsRuntimeExceptionTest extends PmdRuleTst {
```
This class needs an accompanying XML file, that describes the test cases. The file should live under `pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/SignatureDeclareThrowsRuntimeExceptionTest`.
See also https://pmd.github.io/pmd-6.12.0/pmd_userdocs_extending_testing.html

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/SignatureDeclareThrowsRuntimeExceptionRule.java:
```diff
@@ -0,0 +1,81 @@
+     *            to evaluate
+     * @return true if <code>Exception</code> is declared and has proper parents
+     */
+    private boolean hasDeclaredRunTimeExceptionInSignature(ASTName exception) {
```
I'd name this method simply "isRuntimeException" - since the method as no knowledge, that the exception name is used in a method/constructor signature

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/SignatureDeclareThrowsRuntimeExceptionRule.java:
```diff
@@ -0,0 +1,81 @@
+ * @author <a href="prakashbp2020@gmail.com">Bhanu Prakash Pamidi</a>
+ */
+public class SignatureDeclareThrowsRuntimeExceptionRule extends AbstractJavaRule {
+
```
Since this rule only visits two AST nodes, we can make use of the "RuleChain" mechanism. This means, that not the whole AST tree is traversed and visited, but only specific AST nodes. This speeds up the rule execution time.

You declare the nodes, you want to have considered, in the default constructor of your rule:

```java
    public SignatureDeclareThrowsRuntimeExceptionRule() {
        addRuleChainVisit(ASTMethodDeclaration.class);
        addRuleChainVisit(ASTConstructorDeclaration.class);
    }
```

Then, when implementing the visit method, don't call super - since we don't want to traverse the AST anymore:

```java
    @Override
    public Object visit(ASTMethodDeclaration methodDeclaration, Object o) {
        checkExceptions(methodDeclaration, o);
        return o;
    }
```

pamidi99 commented:
Thanks for your feedback @adangel! committed test cases file and let me know if it needs any change. 

Also, I have tried the Rule Designer to implement the Rule via XPath, wanted to see it generate AST for the code but I didn't find any menu option to trigger generation of AST after the code is pasted in source code window of it. Is there any detailed document or wiki for using Rule Designer?

adangel commented on pmd-java/src/main/resources/category/java/design.xml:
```diff
@@ -1136,6 +1136,26 @@ derived from RuntimeException or a checked exception.
+        <priority>3</priority>
+        <example>
+<![CDATA[
+import java.lang.RuntimeException;
```
You do not need to import classes from package `java.lang`. They are automatically imported. PMD actually has a rule to avoid this: [DontImportJavaLang](https://pmd.github.io/pmd-6.12.0/pmd_rules_java_codestyle.html#dontimportjavalang)

adangel commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/SignatureDeclareThrowsRuntimeException.xml:
```diff
@@ -0,0 +1,82 @@
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
+    <test-code>
+        <description><![CDATA[
```
Please remove the surround CDATA section - it makes the XML file much more harder to read

adangel commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/SignatureDeclareThrowsRuntimeException.xml:
```diff
@@ -0,0 +1,82 @@
+     ]]></description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+import java.lang.RuntimeException;
```
Please remove the import for java.lang.Runtime

adangel commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/SignatureDeclareThrowsRuntimeException.xml:
```diff
@@ -0,0 +1,82 @@
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+constructor throws RuntimeException
```
I'd add one additional test case: "constructor throws checked exception", to verify, we don't report a violation on that.

adangel commented:
There is no comprehensive documentation yet about the designer. But
* starting it is mentioned here: https://pmd.github.io/pmd-6.12.0/pmd_userdocs_installation.html
* using it is here: https://pmd.github.io/pmd-6.12.0/pmd_userdocs_extending_writing_xpath_rules.html#using-designer

The designer automatically parses the source code (if it is valid). There is no extra button to trigger this.

I've created a screencast:
![designer-signaturedeclarethrowsruntimeexceptionrule](https://user-images.githubusercontent.com/1573684/53693097-8ad5bf80-3d9b-11e9-9aa4-3bac801e916b.gif)

Note: the XPath expression is not complete yet, e.g. it doesn't consider constructors.

pamidi99 commented:
@adangel Thanks for taking time to create a screenshot of rule designer, appreciate it! 

I have turned the rule to XPath and added a few more test cases to make sure that rule works the same for constructor, class method and interface method and addressed your comments as well. Let me know if this PR still needs any change.

oowekyala commented on pmd-java/src/main/resources/category/java/design.xml:
```diff
@@ -1136,6 +1136,36 @@ derived from RuntimeException or a checked exception.
+        </example>
+    </rule>
+	
+	<rule name="SignatureDeclareThrowsRuntimeException"
```
Maybe the rule should be named `ThrowsClauseDeclaresUncheckedExceptions`, or `AvoidUncheckedExceptionsInSignatures`? IMHO the two verbs glued together (`DeclareThrows`) are very confusing on first read

I think consistency with `SignatureDeclareThrowsException`  can be overlooked, especially since it should be "declares" with an "s" anyway.

@pamidi99 Wdyt?

pamidi99 replied:
I think `AvoidUncheckedExceptionsInSignatures ` looks good to me. @adangel  Wdyt? Once you both confirm the renaming, will make the change.

adangel replied:
sounds good :+1:

oowekyala commented on pmd-java/src/main/resources/category/java/design.xml:
```diff
@@ -1136,6 +1136,36 @@ derived from RuntimeException or a checked exception.
+	<rule name="SignatureDeclareThrowsRuntimeException"
+          since="6.13.0"
+          language="java"
+          message="A method/constructor should not explicitly declare java.lang.RuntimeException or it's subclasses in throws clause of it's signature"
```
```suggestion
          message="A method/constructor should not explicitly declare unchecked exceptions in its ''throws'' clause"
```

oowekyala commented on pmd-java/src/main/resources/category/java/design.xml:
```diff
@@ -1136,6 +1136,36 @@ derived from RuntimeException or a checked exception.
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_design.html#signaturedeclarethrowsruntimeexception">
+        <description>
+A method/constructor should not explicitly declare java.lang.RuntimeException or it's subclasses in throws clause of it's signature, since it is advised to avoid declaring unchecked exceptions in method signature.
```
Suggesting a better practice in rule descriptions is generally a good idea:
```suggestion
A method or constructor should not explicitly declare java.lang.RuntimeException 
or its subclasses in its `throws` clause. Java doesn't force the caller to handle the 
exception, so it's unnecessary except for documentation. A better practice is to 
document the exceptional cases with a `@throws` Javadoc tag, which allows being
more descriptive.
